### PR TITLE
[riscv] Update pytorch to version 2.4.0

### PR DESCRIPTION
**Pull request by iarspider**
I will remove unused patches later, once tests pass.

### CODE REVIEW

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test

**Comment by smuzaffar** [comment]
@iarspider , please check the pytorch logs, looks like ROCM was not enabled properly.  Either see the package doc on what requires to enable it or go through the cmake configuration ... may be we are missing some env to rocm distribution?
```
-- Looking for cgesdd_
-- Looking for cgesdd_ - found
-- Found a library with LAPACK API (open).
disabling ROCM because NOT USE_ROCM is set
-- MIOpen not found. Compiling without MIOpen support
disabling MKLDNN because USE_MKLDNN is not set
-- Looking for clock_gettime in rt
-- Looking for clock_gettime in rt - found
```

**Comment by iarspider** [comment]
Looks like you really can't use both CUDA and ROCm at the same time: https://github.com/pytorch/pytorch/blob/v2.3.1/aten/CMakeLists.txt#L71-L73 .

**Comment by smuzaffar** [comment]
> Looks like you really can't use both CUDA and ROCm at the same time: https://github.com/pytorch/pytorch/blob/v2.3.1/aten/CMakeLists.txt#L71-L73 .

ah ok. By the way, for this build you have both CUDA and ROCM ON, so any idea why cmake did not fail with this error 
```
message(FATAL_ERROR "Both CUDA and ROCm are enabled and found. PyTorch can only be built with either of them. Please turn one off by using either USE_CUDA=OFF or USE_ROCM=OFF.")
```

something must have disabled rocm before cmake tried to configure `aten`

**Comment by iarspider** [comment]
> any idea why cmake did not fail with this error

Yes - it didn't find our installation of ROCm (there are at least two environment variables that need to be set)

**Comment by iarspider** [comment]
please test

**Comment by smuzaffar** [comment]
@iarspider , can you please open a separate PR to build pytorch with ROCM (i.e. disable CUDA and enable ROCM) ?

**Comment by iarspider** [comment]
@smuzaffar will do.

**Comment by iarspider** [comment]
I will test the latest changes locally, in case there are more missing braces. Will retrigger tests once cmssw compiles.

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
Previous build failure not related to this PR.

**Comment by iarspider** [comment]
please test with cms-sw/cmssw#45467

**Comment by smuzaffar** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45467

**Comment by iarspider** [comment]
please abort

**Comment by iarspider** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45467, https://github.com/cms-sw/cmssw/pull/45428

**Comment by iarspider** [comment]
Ok, github doesn't properly render two PR links separated by comma...

**Comment by smuzaffar** [comment]
> Ok, github doesn't properly render two PR links separated by comma...

yes you need to use `,<space>`

**Comment by iarspider** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45467, https://github.com/cms-sw/cmssw/pull/45428
Unit test failures are unrelated

**Comment by smuzaffar** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45467

File: pytorch-scatter.spec [hunk 1]
```diff
@@ -9,14 +10,22 @@ Source: git+https://github.com/%{github_user}/pytorch_scatter.git?obj=%{branch}/
...
 # Make sure the default c++sdt stand is c++14
 grep -q 'CMAKE_CXX_STANDARD  *14' CMakeLists.txt
 sed -i -e 's|CMAKE_CXX_STANDARD  *14|CMAKE_CXX_STANDARD %{cms_cxx_standard}|' CMakeLists.txt
 
+USE_CUDA=OFF
+%if "%{cmsos}" != "slc7_aarch64"
+if [ "%{cuda_gcc_support}" = "true" ] ; then
+USE_CUDA=%{!?without_cuda:ON}
+fi
+%endif
```
Comment (smuzaffar) [inline pytorch-scatter.spec:27]: @iarspider , can you please update all these `pytorch-*.spec` ( where needed) and use something like
```
USE_CUDA=OFF
%if 0%{!?without_cuda:1}
if [ "%{cuda_gcc_support}" = "true" ] ; then
USE_CUDA=ON
fi
%endif
```

**Comment by iarspider** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45467

**Comment by iarspider** [comment]
please test with https://github.com/cms-sw/cmssw/pull/45467

**Comment by smuzaffar** [comment]
test parameters:
- pull_request = https://github.com/cms-sw/cmssw/pull/45467

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test for el8_aarch64_gcc12

**Comment by smuzaffar** [comment]
please test for el9_amd64_gcc12

**Comment by smuzaffar** [comment]
please test for el8_aarch64_gcc12

**Comment by smuzaffar** [comment]
please test for el9_amd64_gcc13

**Comment by smuzaffar** [comment]
+externals

looks good

**Pull request merged**